test/distributed/test_distributed.py

```diff
@@ -2189,6 +2190,40 @@ def test_DistributedDataParallel_SyncBatchNorm_Diff_Input_Sizes_gradient(self):
             global_bs=global_bs,
             offset=bs_offset)
 
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_gpu
+    def test_DistributedDataParallel_SyncBatchNorm_Empty_Batch_Size(self):
+        group, group_id, rank = self._init_global_test()
+        # only do single GPU per process
+        gpus = [rank]
+
+        # cpu training setup
+        model = BN_NET
+
+        num_processes = int(WORLD_SIZE)
+
+        # local batch sizes
+        bs_array = torch.randint(low=0, high=10, size=(num_processes,), dtype=torch.int)
+        if (bs_array != 0).all().item():
+            bs_array[0] = 0     # make sure at lease one local batch is empty
+            if num_processes > 2:
+                bs_array[num_processes // 2] = 0    # another empty batch in the middle
+        if bs_array.sum().item() == 0:
+            bs_array[0] = 1     # make sure the batch is not fully empty
+
+        local_bs = bs_array[rank].item()
+        bs_offset = bs_array[:rank].sum().item()
+        global_bs = bs_array.sum().item()
+
+        self._test_DistributedDataParallel_SyncBatchNorm(
```

Comment:
this test does not test that batch norm parameters will get non-None gradients.

Reply:
I think the test asserts the parameter between DDP model and base model after gradient have been applied: https://github.com/pytorch/pytorch/blob/5be1ee032211b6bc7a271ea2e4c1ca4d1b1fe0d0/test/distributed/test_distributed.py#L1893-L1895
non-None gradient is a required other-wise parameters would differ during training.

Reply:
I don't think so. _assert_equal_param compares parameters only. You expect gradients to be zero for empty batches. If they are None instead of zero, they won't be applied here https://github.com/pytorch/pytorch/blob/5be1ee032211b6bc7a271ea2e4c1ca4d1b1fe0d0/test/distributed/test_distributed.py#L1843-L1848 and parameter comparison would still succeed.

Reply:
I stand corrected. :stuck_out_tongue: 
@xwang233 we need to update the test.